xarray/core/treenode.py

```diff
@@ -333,9 +333,9 @@ def subtree(self: Tree) -> Iterator[Tree]:
         --------
         DataTree.descendants
         """
-        from xarray.datatree_.datatree import iterators
+        from xarray.core.iterators import PreOrderIter
 
-        return iterators.PreOrderIter(self)
+        return PreOrderIter(self)
```

Comment:
So we currently have two patterns of iteration used: depth-first (`PreOrderIter`) and breadth-first (`LevelOrderIter`). It looks like `PreOrderIter` is used in `TreeNode.subtree`, and `LevelOrderIter` is used in `mapping.diff_treestructure`. `diff_treestructure` is called inside `check_isomorphic`, and every other time we iterate over the tree it just calls `.subtree`.

Why the distinction? In `diff_treestructure` when comparing two trees with multiple points at which their structure deviates, the order of iteration will affect which deviation is raised as an error first. @flamingbear and I think here is it more intuitive to raise errors from the top-level of the tree first, so we agree that `LevelOrderIter` is the right choice here.

For mapping over the nodes of the tree in `.subtree`, I had previously thought that there was a good reason to use `PreOrderIter`. My reasoning was to do with how `dt['/a/b/c/d/'] = data` would immediately create the nodes `/a`, `/a/b`, `/a/b/c` even if they didn't already exist. But actually I don't think that's relevant here.

Another reason to use `PreOrderIter` might be around using `map_over_subtree` to map an operation that then fails. For example taking `dt.mean(dim='time')` when some of the nodes doesn't have a time dimension. Again the order of iteration affects which node will raise the first error.

However, if we actually think that `LevelOrderIter` is fine for the `.subtree` case too, we can potentially simplify this PR considerably by replacing the whole `AbstractIter/PreOrderIter/LevelOrderIter` framework with just a single `iterate_breadth_first` function that returns an iterator, and use that in all cases.

Reply:
(Also then we wouldn't need the anytree license)

Reply:
@TomNicholas - I've made some updates (or rather nabbed some work @flamingbear did and made some `mypy` tweaks).

A couple of things here:

* This now contains the single class, but hasn't wrapped it in a function. I think it gets at what you were after, but just FYI. (Although, part of me wonders if this now warrants a module all of it's own?)
* I left the `anytree` license in the PR, because the `LevelOrderIter` is still 99% the code from `anytree`.

Reply:
> This now contains the single class, but hasn't wrapped it in a function.

That's fine, thanks.

> (Although, part of me wonders if this now warrants a module all of it's own?)

If you wanted to move it to `treenode.py` instead that would also make sense. But I don't have a strong opinion, and it's trivial to change later.

Reply:
Okay - I'm happy to leave this in it's own module for now.